### PR TITLE
rpm: fix build on Darwin

### PR DESCRIPTION
**Pull request by reckenrode**
RPM needs a 13.x SDK or newer.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by emilazy** [comment]
Finally, macOS Nix users can have a decent package manager.

File: pkgs/tools/package-management/rpm/default.nix [hunk 1]
```diff
@@ -34,13 +37,33 @@ stdenv.mkDerivation rec {
     hash = "sha256-N/O0LAlmlB4q0/EP3jY5gkplkdBxl7qP0IacoHeeH1Y=";
   };
 
+  patches = [
+    # Resolves `error: expected expression` on clang
+    # See: https://github.com/rpm-software-management/rpm/issues/2435.
+    (fetchpatch2 {
+      url = "https://github.com/rpm-software-management/rpm/commit/3b7cc5b5362fc22cb6f96180741e2701e703844d.diff?full_index=1";
+      hash = "sha256-pBKYQK97RkK9bigroNVgeDrR37poFSUBdEpB/RjwK5A=";
+    })
```
Comment (emilazy) [inline pkgs/tools/package-management/rpm/default.nix:46]: This is included in 4.19 and 4.20; we should probably bump the version instead. If not, we should at least point to `b960c0b43a080287a7c13533eeb2d9f288db1414`, which is the version of the commit that was landed and won’t get GC’d.

File: pkgs/tools/package-management/rpm/default.nix [hunk 2]
```diff
@@ -34,13 +37,33 @@ stdenv.mkDerivation rec {
...
+    # Fix missing includes required to build on Darwin.
+    # See: https://github.com/rpm-software-management/rpm/pull/2571.
+    (fetchpatch2 {
+      url = "https://github.com/rpm-software-management/rpm/commit/f07875392a09228b1a25c1763a50bbbd0f6798c2.diff?full_index=1";
+      hash = "sha256-DLpzMApRCgI9zqheglFtqL8E1vq9X/aQa0HMnIAQgk8=";
+    })
+    (fetchpatch2 {
+      url = "https://github.com/rpm-software-management/rpm/commit/b2e67642fd8cb64d8cb1cca9e759396c1c10807d.diff?full_index=1";
+      hash = "sha256-q3fIBfiUJVmw6Vi2/Bo/zX6/cqTM7aFnskKfMVK3DlU=";
+    })
```
Comment (emilazy) [inline pkgs/tools/package-management/rpm/default.nix:56]: This PR was also rebased for merge and is in 4.20.

**Review by emilazy — CHANGES_REQUESTED** [review]
Could you try bumping to 4.20 instead, which includes these fixes? Otherwise we need to adjust the commit hashes.

File: pkgs/tools/package-management/rpm/default.nix [hunk 3]
```diff
@@ -92,9 +115,6 @@ stdenv.mkDerivation rec {
     license = with licenses; [ gpl2Plus lgpl21Plus ];
     description = "RPM Package Manager";
     maintainers = with maintainers; [ copumpkin ];
-    platforms = platforms.linux;
-    # Support for darwin was removed in https://github.com/NixOS/nixpkgs/pull/196350.
-    # This can be re-enables for apple_sdk.version >= 13.0.
-    badPlatforms = platforms.darwin;
+    platforms = platforms.linux ++ platforms.darwin;
```
Comment (emilazy) [inline pkgs/tools/package-management/rpm/default.nix:118]: BTW, I’m guessing this can be `platforms.unix`.

**Comment by reckenrode** [comment]
RPM 4.20 is a non-trivial effort to package. That’s way beyond what I want to do to update it for Darwin. If this isn’t mergable, I’d rather leave Darwin marked broken.

**Comment by emilazy** [comment]
I think it’s fine to merge this and let someone else handle RPM 4.20 as long as we can point the fetches to the copies of the commits that were actually merged rather than the potentially ephemeral ones from the PRs.

**Comment by reckenrode** [comment]
I fixed the commit to be the one from the repo.

**Comment by emilazy** [comment]
The other PR needs its commits changing too (it seems like they do rebase merges).

**Review by paparodeo — APPROVED** [review]

**Comment by reckenrode** [comment]
No changes except I retargeted to master due to the low number of rebuilds.

**Comment by reckenrode** [comment]
@emilazy Is this good to merge? It’s blocked currently.

**Review by emilazy — APPROVED** [review]
Forgot I had a blocking review on this, sorry. Seems fine, hopefully someone else can figure out the new build system another time.

**Pull request merged**